[#28] 스케줄 기능 개발

kylekim2123 opened the pull request:
## 👨‍💻 작업 사항

<!-- 이슈 번호 태그 -->
> **이슈 #28**

### 📑 PR 개요
<!-- PR에 대한 간단한 개요를 작성 -->
- 스케줄 생성, 삭제 기능 개발 및 테스트 코드 작성
- 스케줄 생성, 삭제 관련 예외 처리

---

<!-- 이슈에서 지정했던 작업 목록의 완료 상태를 표시 -->
### ✅ 작업 목록
- [x] 스케줄 생성 기능
  - [x] 공연장 좌석 수 조회 기능
- [x] 스케줄 삭제 기능
  - [x] 스케줄 존재 여부 확인 기능
- [ ] 예외 처리
  - [ ] 생성 시, performanceId에 해당하는 공연이 없는 경우 예외 -> `공연 기능 완성되면 예외처리 할 예정`
  - [x] 생성 시, startDatetime 이 "yyyy-MM-dd HH:mm:ss" 형식에 맞지 않게 입력되면 예외
  - [ ] 생성 시, startDatetime 이 해당 공연의 시작~종료 날짜 사이에 없는 경우 예외 -> `공연 기능 완성되면 예외처리 할 예정`
  - [x] 생성 시, sequence 가 1이상의 자연수가 아닌 경우 예외
  - [x] 삭제 시, performanceId와 ScheduleId에 해당하는 스케줄이 없는 경우 예외
- [x] 스케줄 테스트 코드 작성
  - [x] 스케줄 데이터를 DB에 삽입 한 후 ID 반환 테스트
  - [x] 공연 ID를 이용한, 공연이 진행되는 공연장의 좌석 수 조회 테스트
  - [x] 특정 스케줄 ID에 해당하는 스케줄이 존재하는지 아닌지 판별 테스트
  - [x] 특정 스케줄 ID에 해당하는 스케줄을 삭제 테스트

---

### 🙏 리뷰어에게
<!-- PR에 작성한 변경 사항에 대해 팀원들에게 설명 -->
- 아직 진행하지 못한 예외 처리는 공연 기능 완성 후에 할 예정입니다.
- 테스트 코드를 많이 작성 안해봐서 개인적으로 좀 아쉽습니다. 테스트 코드 관련 지적 환영합니다!

---

### Prefix
> PR 코멘트를 작성할 때 항상 Prefix를 붙여주세요.
- `P1`: 꼭 반영해주세요 (Request changes)
- `P2`: 적극적으로 고려해주세요 (Request changes)
- `P3`: 웬만하면 반영해 주세요 (Comment)
- `P4`: 반영해도 좋고 넘어가도 좋습니다 (Approve)
- `P5`: 그냥 사소한 의견입니다 (Approve)

kylekim2123 committed:
feat: 스케줄 생성 관련 repository 기능 추가

kylekim2123 committed:
test: 스케줄 생성 관련 repository 기능 테스트 코드 추가

kylekim2123 committed:
feat: 스케줄 생성 관련 service 기능 추가

kylekim2123 committed:
test: 공연장 좌석 수 조회 기능 테스트 코드 추가

kylekim2123 committed:
feat: 스케줄 생성 관련 controller 기능 추가

kylekim2123 committed:
feat: 스케줄 삭제 관련 repository 기능 추가

kylekim2123 committed:
test: 스케줄 삭제 관련 repository 기능 테스트 코드 추가

kylekim2123 committed:
refactor: 스케줄 삭제 기능 예외처리
- 존재하지 않는 공연 ID와 스케줄 ID의 스케줄을 삭제하는 경우

kylekim2123 committed:
test: 스케줄 상세 조회 기능 생성 및 저장 로직 테스트 코드 변경

kylekim2123 committed:
feat: 스케줄 요청 DTO 필드에 Bean Validation 적용
- startDatetime : "yyyy-MM-dd HH:mm:ss" 형식 여부 검증
- sequence : 1이상의 자연수 여부 검증

hanjo8813 commented on src/main/java/com/programmers/ticketparis/controller/schedule/ScheduleController.java:
```diff
@@ -0,0 +1,42 @@
+
+    @PostMapping("/schedules")
+    @ResponseStatus(HttpStatus.CREATED)
+    public ApiResponse<ScheduleResponse> createSchedule(@PathVariable Long performanceId, @Valid @RequestBody ScheduleCreateRequest request, HttpServletRequest httpServletRequest) {
```
`p4`
파라미터가 너무 길어지면 한눈에 보기 힘들 수 있으니 적절히 띄워주면 좋을것 같습니다.
```suggestion
    public ApiResponse<ScheduleResponse> createSchedule(
        @PathVariable Long performanceId, 
        @Valid @RequestBody ScheduleCreateRequest request, HttpServletRequest httpServletRequest
    ) {
```

추가로 hard wrap에 대한 code style 컨벤션도 정했나요?

kylekim2123 replied:
반영: [313ac9a](https://github.com/prgrms-be-devcourse/BE-04-TICKETPARIS/pull/30/commits/313ac9afb4503e98fe2d153024cda2aff1dafe32)
- hard wrap을 120으로 가기로 다같이 정했습니다. 관련해서 수정도 했습니다!

hanjo8813 commented on src/main/java/com/programmers/ticketparis/service/schedule/ScheduleService.java:
```diff
@@ -0,0 +1,51 @@
+    @Transactional
+    public ScheduleResponse createSchedule(Long performanceId, ScheduleCreateRequest request) {
+        // performanceId에 해당하는 공연 데이터가 없는 경우 예외 처리 예정 (2023.09.06 김영주 작성)
+        // startDatetime이 공연의 시작, 종료날짜 범위를 벗어나는 경우 예외 처리 예정 (2023.09.06 김영주 작성)
```
`p4`
TODO 주석을 사용해보세요~

```suggestion
        // TODO: performanceId에 해당하는 공연 데이터가 없는 경우 예외 처리 예정 (2023.09.06 김영주 작성)
        // TODO: startDatetime이 공연의 시작, 종료날짜 범위를 벗어나는 경우 예외 처리 예정 (2023.09.06 김영주 작성)
```

참고 : https://cheese10yun.github.io/intellij-todo/

kylekim2123 replied:
반영: [313ac9a](https://github.com/prgrms-be-devcourse/BE-04-TICKETPARIS/pull/30/commits/313ac9afb4503e98fe2d153024cda2aff1dafe32)
- 이런 기능이 있는 줄 몰랐는데 정말 유용하네요! 완전 감사합니다 ㅎㅎ

hanjo8813 reviewed (commented):
다른 코드는 최대한 개입하지 않고, 추천할만한 사항에 대해서만 간단히 리뷰를 남겼습니다!
다른 팀원분들 리뷰가 끝난 후 개선점이 남아있다면 추가로 코멘트 남길게요~

Hchanghyeon commented on src/main/resources/mapper/ScheduleMapper.xml:
```diff
@@ -0,0 +1,33 @@
+          AND schedule_id = #{scheduleId};
+    </select>
+
+    <delete id="deleteById">
+        DELETE
+        FROM schedule
+        WHERE performance_id = #{performanceId}
+          AND schedule_id = #{scheduleId};
+    </delete>
```
`p4`
클라이언트에게 전달 받을 때 scheduleId만 받아서 삭제도 가능할 것 같은데 performanceId까지 받아서 처리하시는 이유가 궁금합니다!

kylekim2123 replied:
구두로 말씀드렸지만 서면으로도 말씀드립니다!
Rest API에서 URI는 자원을 표기하기에, 
스케줄의 삭제는 `DELETE /api/performances/{performanceId}/schedules/{scheduleId}` 로 표기하고 있습니다.
따라서 단순히 `몇 번 스케줄을 삭제한다.` 보다는 `몇 번 공연의 몇 번 스케줄을 삭제한다.` 라는 의미가 강하다고 저는 생각합니다.
그래서 공연 ID와 스케줄 ID를 같이 받아서 삭제한다고 보시면 될 것 같습니다.

Hchanghyeon commented on src/main/java/com/programmers/ticketparis/service/schedule/ScheduleService.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Transactional
+    public void deleteScheduleById(Long performanceId, Long scheduleId) {
+        Integer deletedCounts = scheduleRepository.deleteById(performanceId, scheduleId);
```
`p4`
deleteById를 진행하기 전에 미리 existById로 존재하는지 체크하는건 어떻게 생각하시나요? 

저의 경우 JPA에도 deleteById를 하면 findById + delete가 합쳐지듯 MyBatis에서도 어떠한 동작을 할 때 미리 값이 있는지 체크해보아야한다고 생각이 들어서 영주님의 생각이 궁금합니다!

charlesuu replied:
p5
 "deleteById를 진행하기 전에 existById로 값이 존재하는지 체크하는 행위" 를 통해 얻을 수 있는 이점은, 값이 존재하지 않을 때 원인에 대한 정보를 담은 명확한 비즈니스 예외를 던질 수 있게 된다는 점이라고 생각합니다. 위 코드는 비즈니스 예외를 만드는 과정은 다르지만 결론적으로는 existById로 체크하는 것과 같은 결과를 내고 있다는 생각도 드는데 다른 분들은 어떻게 생각하시나요??

kylekim2123 replied:
반영: [b3031c6](https://github.com/prgrms-be-devcourse/BE-04-TICKETPARIS/pull/30/commits/b3031c684f87f4839290e4adbd247ff41a119fb6)
- 기존의 deletedCounts를 이용해서 exists 여부를 검증했던 코드를 삭제했습니다.
- 별도의 existsById를 만들어서, 삭제 전 존재 여부를 검증하도록 했습니다.

Hchanghyeon commented on src/main/java/com/programmers/ticketparis/controller/schedule/ScheduleController.java:
```diff
@@ -0,0 +1,42 @@
+    @ResponseStatus(HttpStatus.CREATED)
+    public ApiResponse<ScheduleResponse> createSchedule(@PathVariable Long performanceId, @Valid @RequestBody ScheduleCreateRequest request, HttpServletRequest httpServletRequest) {
+        ScheduleResponse response = scheduleService.createSchedule(performanceId, request);
+        String requestURI = httpServletRequest.getRequestURI();
+
+        return ApiResponse.of(requestURI, response);
```
`p3`
ApiResponse에서 path 정보를 클라이언트에게 보내주기 위한 코드 같은데 아무래도 응답 값이 있는 모든 메서드에 적어줘야하다보니 최대한 코드를 줄이는 것도 좋을 것 같습니다! 아래와 같이 ApiResponse에는 httpServletRequest를 매개변수로 넘겨주고 ApiResponse에서 getRequestURI로 가져와서 처리하는 것은 어떻게 생각하시나요?

```suggestion
        return ApiResponse.of(httpServletRequest, response);
```

mentor-tyler replied:
메소드 시그니쳐만 봐도 어떤행위를 하는지 알수 있어야 한다고 생각하는데요.
httpServletRequest 안에 무엇을 가져다 쓰지? 하면 해당 of method를 다시 들어가봐야 하지 않을까 싶기도 합니다.
다만, 지금의 코드에서는 간단한 코드라 아래처럼 조정하면 어떨지 싶네요.

\+ 좀 더 하자면 파라미터가 2개 이상일 경우엔 of보다 빌더가 좋지 앟을까 하는 생각도 들어요.
public XXX of(String a, String b) 이렇게 되면 a 와 b를 바꿔 사용해도 컴파일에선 에러가 안나거등요. 물론 빌더도 마찬가지지만 파라미터를 지정해서 셋팅 해주니 좀더 명확하지 않을까 싶기도 하고요 ㅎ

```java
return ApiResponse.of(httpServletRequest.getRequestURI(), response);
```

hanjo8813 replied:
이런 번거로운 점 때문에 `ResponseBodyAdvice`를 쓰는것 같습니다.
도입을 검토해보세요~!~!~!

Hchanghyeon commented on src/main/java/com/programmers/ticketparis/dto/schedule/request/ScheduleCreateRequest.java:
```diff
@@ -0,0 +1,31 @@
+    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
+    private LocalDateTime startDatetime;
+
+    @Positive
```
`p2`
`@Postive`로 인한 예외가 발생했을 때 예외에 대한 설명(message)이 있으면 더 좋을 것 같습니다!

kylekim2123 replied:
반영: [72509ee](https://github.com/prgrms-be-devcourse/BE-04-TICKETPARIS/pull/30/commits/72509eebf5d93760fd96b97fe2267671cabeddc1)
- 말씀하신 대로 예외 메시지를 추가했습니다.
- Bean Validation 예외 메시지의 로깅 처리는, 추후에 별도의 PR을 통해 일괄적으로 수정하는게 좋을 것 같아서 일단 하지 않았습니다.

Hchanghyeon commented on src/main/java/com/programmers/ticketparis/exception/GlobalExceptionHandler.java:
```diff
@@ -56,6 +58,22 @@ public ResponseEntity<ErrorResponse> handleMethodArgumentNotValidException(Metho
+        }
+
+        return ResponseEntity.status(BAD_REQUEST.getStatus()).body(response);
+    }
```
`p5`
너무 좋아요 👍👍👍👍👍👍👍

Hchanghyeon reviewed (commented):
고생하셨습니다 영주님~! 👍👍

궁금한 사항들 리뷰 달아놓았습니다!  확인 부탁드리겠습니다!! 🙏

mentor-tyler commented on src/main/java/com/programmers/ticketparis/controller/schedule/ScheduleController.java:
```diff
@@ -0,0 +1,42 @@
+package com.programmers.ticketparis.controller.schedule;
```
`P3`

com.programmers.ticketparis.schedule.controller ㄱㅏ 아닌
com.programmers.ticketparis.controller.schedule 로 하셨군요~

크게 상관은 없지만 보통 멀티모듈(확장성)까지 고려해서 pacakge 를 도메인 단위로 나누곤 합니다~
어느 package 하위가 무거워 지는 날이 올테고 그때 멀티모듈을 고려해볼수 있는 근거가 되어서요~

kylekim2123 replied:
안그래도 프로젝트 설계 단계에서 팀원들끼리 두 옵션 중에 어떤 걸로 할지 고민을 많이 했었던 것 같습니다.
MyBatis의 type-aliases-package 옵션을 편하게 쓰기 위해서, 상위 패키지 내에 여러 도메인 타입 정보를 몰아넣게 되었는데요.
멘토님 말씀 듣고 나니까, 확장성을 고려하면 도메인 별로 나누는 것이 더 좋을 것 같습니다!
팀원들과 상의 해보고, 해당 사항은 별도의 PR로 한번에 바꿔보도록 하겠습니다.

mentor-tyler commented on src/main/java/com/programmers/ticketparis/repository/schedule/MybatisScheduleRepository.java:
```diff
@@ -0,0 +1,38 @@
+
+@Repository
+@RequiredArgsConstructor
+public class MybatisScheduleRepository implements ScheduleRepository {
```
다른 ORM을 고려하신 네이밍이군요?

kylekim2123 replied:
넵 맞습니다! 그런데 해당 프로젝트에서는 MyBatis만 사용하게 될 가능성이 높아서 팀원들이랑 상의 해보고 맞춰야 할 것 같습니다!

mentor-tyler commented on src/main/resources/mapper/ScheduleMapper.xml:
```diff
@@ -0,0 +1,33 @@
+    </insert>
+
+    <select id="findHallSeatsCountByPerformanceId" resultType="Integer">
+        SELECT B.seats_count
+        FROM performance as A
+                 JOIN hall as B
+                      ON A.hall_id = B.hall_id
+        WHERE A.performance_id = #{performanceId};
```
테이블 alias를 사용할땐 최소한 alias되는 키워드만 봐도 무슨 테이블인지 유추가 가능했음 좋겠어요.

지금의 경우... A 보단 p, B 보단 h

kylekim2123 replied:
반영: [313ac9a](https://github.com/prgrms-be-devcourse/BE-04-TICKETPARIS/pull/30/commits/313ac9afb4503e98fe2d153024cda2aff1dafe32)
- 너무 좋습니다! 말씀 하신 대로 변경했습니다. 앞으로도 이런 점은 계속 주의해야겠네요!

mentor-tyler reviewed (commented):
👍

kylekim2123 committed:
style: 개행, SQL, 주석 관련 수정
- Hard Wrap 옵션 200 -> 120 변경으로 인한 개행 추가
- SELECT 문에서 * 대신 컬럼 명시
- 할일 주석에 대해 TODO 키워드 추가

kylekim2123 committed:
refactor: 스케줄 삭제 시, 해당 스케줄이 존재하는지 먼저 확인 후 삭제하도록 변경

kylekim2123 committed:
test: 스케줄 조회와 존재 여부 판별 기능에 대한 테스트 코드 작성

charlesuu commented on src/main/java/com/programmers/ticketparis/repository/schedule/MybatisScheduleRepository.java:
```diff
@@ -0,0 +1,38 @@
+
+@Repository
+@RequiredArgsConstructor
+public class MybatisScheduleRepository implements ScheduleRepository {
```
p5
생각해보니 ORM을 확장할 계획이 없다면 네이밍을 단순화하는 것도 좋을 것 같아요~ 영주님은 어떻게 생각하시나요?

kylekim2123 replied:
네 저도 현호님 말씀에 동감해요! 이건 저희끼리 합의에 의해서 맞춰봐야 할 것 같네요~!

charlesuu commented on src/test/java/com/programmers/ticketparis/service/schedule/ScheduleServiceTest.java:
```diff
@@ -0,0 +1,29 @@
+    void test_delete_schedule_throw_exception(Long performanceId, Long scheduleId) {
+        //when, then
+        assertThatThrownBy(() -> scheduleService.deleteScheduleById(performanceId, scheduleId))
+            .isInstanceOf(ScheduleException.class);
```
p2
.hasMessageContaining() 메서드로 예외의 message 내용이 잘 들어갔는지 검증할 수 있습니다! ExceptionRule의 message가 그대로 들어가고 있으니 이대로 추가 해주셔도 될 것 같아요~!
```suggestion
            .isInstanceOf(ScheduleException.class)
            .hasMessageContaining("공연 ID와 스케줄 ID에 해당하는 스케줄이 존재하지 않아 삭제 불가");
```

kylekim2123 replied:
반영: [b3031c6](https://github.com/prgrms-be-devcourse/BE-04-TICKETPARIS/pull/30/commits/b3031c684f87f4839290e4adbd247ff41a119fb6)
- 너무 좋네요! 다만, 메시지를 하드 코딩 하지는 않고 ExceptionRule에 적힌 내용을 불러왔습니다.
- 추가적으로 hasMessage()는 정확히 예외 메시지가 일치하는지를 보기 때문에 정확성을 위해 Containing 대신 hasMessage를 사용했습니다.

charlesuu reviewed (approved):
영주님 수고하셨습니다~!! 몇 가지 피드백 드렸는데 확인 부탁드려요
전반적으로 다가~~~  lgtm!!

kylekim2123 committed:
refactor: 스케줄 생성 DTO에 검증 메시지 추가

kylekim2123 committed:
fix: 스케줄 시작 날짜/시간이 달력/시계 범위를 넘어가는 경우 예외 처리
